# Clustering Data to Unveil Maji Ndogo's Water Crisis.

## Introduction
In this second part of the integrated project, we gear up for a deep analytical dive into Maji Ndogo's water scenario. Harness the power of a wide range of SQL functions, including intricate window functions, to tease out insights from the data tables.

# Notebook Setup

In [1]:
# Load the sql extension
%load_ext sql

To reload the extension, use reload_ext sql

In [2]:
# Create a connection to the mysql 'md_water_services' databases
%sql mysql+pymysql://root:L0xbysmrben%23@localhost:3306/md_water_services

For connection: `mysql + pymsql://root:password@localhost:port/database`

## Cleaning the Data

Let's bring up the `employee` entity. It has information on all of Maji Ndogo's workers, but note that the email addresses have not been added. If we will have to send them reports and figures, we will need their emails hence we need to update the `email` attribute. Luckily the emails for the organisations, per the project description, are easy: `first_name.last_name@ndogowater.gov`.

We can determine the email address for each employee by:

* selecting the employee_name column

* replacing the space with a full stop

* make it lowercase

* and stitch it all together.


We have to update the database again with these email addresses, so before we do, we can use a `SELECT` query to get the format right, then use `UPDATE` and `SET` to persist the changes into the database.

In [4]:
%%sql
# Construct the email addresses for maji ndogo's workers
SELECT 
	CONCAT(
    LOWER(REPLACE(employee_name, " ", ".")), "@ndogowater.gov") AS new_email
FROM employee
LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
10 rows affected.


[('amara.jengo@ndogowater.gov',),
 ('bello.azibo@ndogowater.gov',),
 ('bakari.iniko@ndogowater.gov',),
 ('malachi.mavuso@ndogowater.gov',),
 ('cheche.buhle@ndogowater.gov',),
 ('zuriel.matembo@ndogowater.gov',),
 ('deka.osumare@ndogowater.gov',),
 ('lalitha.kaburi@ndogowater.gov',),
 ('enitan.zuri@ndogowater.gov',),
 ('farai.nia@ndogowater.gov',)]

In [5]:
%sql SELECT * FROM employee LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[(0, 'Amara Jengo', '+99637993287', 'amara.jengo@ndogowater.gov', '36 Pwani Mchangani Road', 'Sokoto', 'Ilanga', 'Field Surveyor'),
 (1, 'Bello Azibo', '+99643864786', 'bello.azibo@ndogowater.gov', '129 Ziwa La Kioo Road', 'Kilimani', 'Rural', 'Field Surveyor'),
 (2, 'Bakari Iniko', '+99222599041', 'bakari.iniko@ndogowater.gov', '18 Mlima Tazama Avenue', 'Hawassa', 'Rural', 'Field Surveyor'),
 (3, 'Malachi Mavuso', '+99945849900', 'malachi.mavuso@ndogowater.gov', '100 Mogadishu Road', 'Akatsi', 'Lusaka', 'Field Surveyor'),
 (4, 'Cheche Buhle', '+99381679640', 'cheche.buhle@ndogowater.gov', '1 Savanna Street', 'Akatsi', 'Rural', 'Field Surveyor')]

Awesome, now we have emails for all employees persisted in the database. let's check the phone_number entity. The phone numbers should be 12 characters long but as we can see below 👇🏾, the phone numbers are 13 numbers long.

In [6]:
%%sql
# Check the length of the phone numbers
SELECT LENGTH(phone_number) FROM md_water_services.employee LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[(12,), (12,), (12,), (12,), (12,)]

In [7]:
%%sql
# Confirm that the phone_number attribute was updated
SELECT LENGTH(phone_number) FROM md_water_services.employee LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[(12,), (12,), (12,), (12,), (12,)]

## Honoring Employees

Before we can begin the analysis to find employees worth honoring, let's first how many employees reside in each province.

In [8]:
%%sql
SELECT * FROM employee LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
10 rows affected.


[(0, 'Amara Jengo', '+99637993287', 'amara.jengo@ndogowater.gov', '36 Pwani Mchangani Road', 'Sokoto', 'Ilanga', 'Field Surveyor'),
 (1, 'Bello Azibo', '+99643864786', 'bello.azibo@ndogowater.gov', '129 Ziwa La Kioo Road', 'Kilimani', 'Rural', 'Field Surveyor'),
 (2, 'Bakari Iniko', '+99222599041', 'bakari.iniko@ndogowater.gov', '18 Mlima Tazama Avenue', 'Hawassa', 'Rural', 'Field Surveyor'),
 (3, 'Malachi Mavuso', '+99945849900', 'malachi.mavuso@ndogowater.gov', '100 Mogadishu Road', 'Akatsi', 'Lusaka', 'Field Surveyor'),
 (4, 'Cheche Buhle', '+99381679640', 'cheche.buhle@ndogowater.gov', '1 Savanna Street', 'Akatsi', 'Rural', 'Field Surveyor'),
 (5, 'Zuriel Matembo', '+99034075111', 'zuriel.matembo@ndogowater.gov', '26 Bahari Ya Faraja Road', 'Kilimani', 'Rural', 'Field Surveyor'),
 (6, 'Deka Osumare', '+99379364631', 'deka.osumare@ndogowater.gov', '104 Kenyatta Street', 'Akatsi', 'Rural', 'Field Surveyor'),
 (7, 'Lalitha Kaburi', '+99681623240', 'lalitha.kaburi@ndogowater.gov', '145 Sungura Amanpour Road', 'Kilimani', 'Rural', 'Field Surveyor'),
 (8, 'Enitan Zuri', '+99248509202', 'enitan.zuri@ndogowater.gov', '117 Kampala Road', 'Hawassa', 'Zanzibar', 'Field Surveyor'),
 (10, 'Farai Nia', '+99570082739', 'farai.nia@ndogowater.gov', '33 Angélique Kidjo Avenue', 'Amanzi', 'Dahabu', 'Field Surveyor')]

In [14]:
%%sql
# Count the number of employees per province
SELECT province_name, COUNT(employee_name) AS no_of_employees
FROM employee
GROUP BY province_name
ORDER BY no_of_employees DESC;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('Hawassa', 15),
 ('Akatsi', 13),
 ('Kilimani', 12),
 ('Sokoto', 9),
 ('Amanzi', 7)]

Let's check how many employees reside in each town as well.

In [19]:
%%sql 
SELECT town_name, COUNT(employee_name) AS no_of_employees
FROM employee
GROUP BY town_name
ORDER BY no_of_employees DESC;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
9 rows affected.


[('Rural', 29),
 ('Dahabu', 6),
 ('Harare', 5),
 ('Lusaka', 4),
 ('Zanzibar', 4),
 ('Ilanga', 3),
 ('Serowe', 3),
 ('Kintampo', 1),
 ('Yaounde', 1)]

Assuming we asked by those in leadership positions in the organisation to send out an email or message congratulating the top 3 field surveyors. We could use the database to get the `employee_ids` and use those to get the `names`, `email` and `phone_numbers` of the ***three field surveyors with the most location visits**. To do this let's first query the `visits` entity to retrieve the number of visits made by employees.

In [3]:
%%sql 
SELECT * FROM visits 
LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
10 rows affected.


[(0, 'SoIl32582', 'SoIl32582224', datetime.datetime(2021, 1, 1, 9, 10), 1, 15, 12),
 (1, 'KiRu28935', 'KiRu28935224', datetime.datetime(2021, 1, 1, 9, 17), 1, 0, 46),
 (2, 'HaRu19752', 'HaRu19752224', datetime.datetime(2021, 1, 1, 9, 36), 1, 62, 40),
 (3, 'AkLu01628', 'AkLu01628224', datetime.datetime(2021, 1, 1, 9, 53), 1, 0, 1),
 (4, 'AkRu03357', 'AkRu03357224', datetime.datetime(2021, 1, 1, 10, 11), 1, 28, 14),
 (5, 'KiRu29315', 'KiRu29315224', datetime.datetime(2021, 1, 1, 10, 17), 1, 9, 40),
 (6, 'AkRu05234', 'AkRu05234224', datetime.datetime(2021, 1, 1, 10, 18), 1, 0, 30),
 (7, 'KiRu28520', 'KiRu28520224', datetime.datetime(2021, 1, 1, 10, 28), 1, 0, 34),
 (8, 'HaZa21742', 'HaZa21742224', datetime.datetime(2021, 1, 1, 10, 37), 1, 0, 6),
 (9, 'AmDa12214', 'AmDa12214224', datetime.datetime(2021, 1, 1, 10, 58), 1, 0, 36)]

In [24]:
%%sql
# Retrieve three employee ids with the most location visits
SELECT assigned_employee_id, COUNT(assigned_employee_id) AS no_of_visits
FROM visits
GROUP BY assigned_employee_id
ORDER BY no_of_visits DESC
LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[(1, 3708), (30, 3676), (34, 3539), (3, 3420), (10, 3407)]

We note that the three employee ids with the most number of visits from the visits entity are:

 |Employee Id | No of Visits|
|---|---|
|1|3708|
|30|3676|
|34|3539|

Now all that's left is to craft a query that will look up the information of the employees with our retrieved employee ids from the previous queries.

In [34]:
%%sql
SELECT assigned_employee_id, employee_name, phone_number, email
FROM employee
WHERE assigned_employee_id IN(1, 30, 34);

 * mysql+pymysql://root:***@localhost:3306/md_water_services
3 rows affected.


[(1, 'Bello Azibo', '+99643864786', 'bello.azibo@ndogowater.gov'),
 (30, 'Pili Zola', '+99822478933', 'pili.zola@ndogowater.gov'),
 (34, 'Rudo Imani', '+99046972648', 'rudo.imani@ndogowater.gov')]

Awesome, we now have the `employee_name`, `phone_number` and `email` columns of the top dog employees.

## Analysing Locations

Looking at the location table, let's focus on the `province_name`, `town_name` and `location_type` to understand where the water sources are in Maji Ndogo. Let's count the records per `town_name` and then count by `province_name`.

In [42]:
%%sql 
SELECT * FROM location LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('AkHa00000', '2 Addis Ababa Road', 'Akatsi', 'Harare', 'Urban'),
 ('AkHa00001', '10 Addis Ababa Road', 'Akatsi', 'Harare', 'Urban'),
 ('AkHa00002', '9 Addis Ababa Road', 'Akatsi', 'Harare', 'Urban'),
 ('AkHa00003', '139 Addis Ababa Road', 'Akatsi', 'Harare', 'Urban'),
 ('AkHa00004', '17 Addis Ababa Road', 'Akatsi', 'Harare', 'Urban')]

In [53]:
%%sql
SELECT province_name, COUNT(province_name) AS records_per_province
FROM location
GROUP BY province_name
ORDER BY records_per_province DESC;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('Kilimani', 9510),
 ('Akatsi', 8940),
 ('Sokoto', 8220),
 ('Amanzi', 6950),
 ('Hawassa', 6030)]

In [55]:
%%sql
SELECT town_name, COUNT(town_name) AS records_per_town
FROM location
GROUP BY town_name
ORDER BY records_per_town DESC
LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
10 rows affected.


[('Rural', 23740),
 ('Harare', 1650),
 ('Amina', 1090),
 ('Lusaka', 1070),
 ('Mrembo', 990),
 ('Asmara', 930),
 ('Dahabu', 930),
 ('Kintampo', 780),
 ('Ilanga', 780),
 ('Isiqalo', 770)]

From this table, it's pretty clear that most of the water sources in the survey are situated in small rural communities, scattered across Maji Ndogo. If we count the records for each province, most of them have a similar number of sources, so every province is well represented in the survey. Let's create a table that shows the number or records and groups them by both province_name and town_name.

In [59]:
%%sql
# Get the number of records per province and town name
SELECT
    province_name,
    town_name,
    COUNT(location_id) AS records_per_town
FROM md_water_services.location
GROUP BY province_name, town_name
ORDER BY province_name, records_per_town DESC
LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
10 rows affected.


[('Akatsi', 'Rural', 6290),
 ('Akatsi', 'Lusaka', 1070),
 ('Akatsi', 'Harare', 800),
 ('Akatsi', 'Kintampo', 780),
 ('Amanzi', 'Rural', 3100),
 ('Amanzi', 'Asmara', 930),
 ('Amanzi', 'Dahabu', 930),
 ('Amanzi', 'Amina', 670),
 ('Amanzi', 'Pwani', 520),
 ('Amanzi', 'Abidjan', 400)]

These results show us that Maji Ndogo's field surveyors did an excellent job of documenting the status of the country's water crisis. Every province and town has many documented sources. This gives us confidence that the data we have is reliable enough to base our decisions on. This is an insight we can use to communicate data integrity, so let's make a note of that. Let's also check the percentage of location types in the location entity.

In [61]:
%%sql 
SELECT location_type, COUNT(location_type) AS records_per_location
FROM location 
GROUP BY location_type
ORDER BY records_per_location DESC;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
2 rows affected.


[('Rural', 23740), ('Urban', 15910)]

In [67]:
%%sql
SELECT ROUND(((23740/ (23740 + 15910))*100)) AS rural_percentage;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
1 rows affected.


[(Decimal('60'),)]

In [66]:
%%sql
SELECT ROUND(((15910 / (15910 + 23740))*100)) AS Urban_percentage

 * mysql+pymysql://root:***@localhost:3306/md_water_services
1 rows affected.


[(Decimal('40'),)]

So again, what are some of the insights we gained from the location table?
1. 
Our entire country was properly canvassed, and our dataset represents the situation on the grod.   
2. .
60% of our water sources are in rural communities across Maji Ndogo. We need to keep this in mind when we make decisions.

## Diving into Sources

Ok, `water_source` is an entity with lots of stories to tell! We have access to different water source types and the number of people using each source. These are some of the questions we are curious about.

1. How many people did we survey in total?
2. 
How many wells, taps and rivers are there
3. 
How many people share particular types of water sources on avera?
4. ?
How many people are getting water from each type of sourceLeof so
urce?
Let's start with the first question. How many people did we survey in total?

total?

In [69]:
%%sql 
SELECT * FROM  water_source LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
10 rows affected.


[('AkHa00000224', 'tap_in_home', 956),
 ('AkHa00001224', 'tap_in_home_broken', 930),
 ('AkHa00002224', 'tap_in_home_broken', 486),
 ('AkHa00003224', 'well', 364),
 ('AkHa00004224', 'tap_in_home_broken', 942),
 ('AkHa00005224', 'tap_in_home', 736),
 ('AkHa00006224', 'tap_in_home', 882),
 ('AkHa00007224', 'tap_in_home', 554),
 ('AkHa00008224', 'well', 398),
 ('AkHa00009224', 'well', 346)]

In [71]:
%%sql
# Retrieve the total number of people served in Maji Ndogo
SELECT SUM(number_of_people_served) AS total_no_served
FROM water_source;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
1 rows affected.


[(Decimal('27628140'),)]

Let's take care of the second question. How many wells, taps and rivers are there?

In [76]:
%%sql
SELECT type_of_water_source, COUNT(type_of_water_source) AS no_of_water_source
FROM water_source
GROUP BY type_of_water_source
ORDER BY no_of_water_source DESC;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('well', 17383),
 ('tap_in_home', 7265),
 ('tap_in_home_broken', 5856),
 ('shared_tap', 5767),
 ('river', 3379)]

We can see that the well is the most popular type of water source serving the population in Maji Ndogo. How about we investigate the next question about our dataset. How many people share particular types of water sources on average?

In [77]:
%%sql
# Retrieve the number of people that share particular types of water sources on average
SELECT
    type_of_water_source,
    ROUND(AVG(number_of_people_served)) AS avg_no_of_people_served
FROM md_water_services.water_source
GROUP BY type_of_water_source
ORDER BY avg_no_of_people_served DESC;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('shared_tap', Decimal('2071')),
 ('river', Decimal('699')),
 ('tap_in_home_broken', Decimal('649')),
 ('tap_in_home', Decimal('644')),
 ('well', Decimal('279'))]

These results are telling us that 644 people share a `tap_in_home` on average. Does that make sense? No it doesn't, right? Remember the few important things that apply to `tap_in_home` and `broken_tap_in_home`? The surveyors combined the data of many households together and added this as a single tap record, but each household actually has its own tap. In addition to this, there is an average of 6 people living in a home. So 6 people actually share 1 tap (not 644). We can see that the `shared_tap` has the highest number of `total_no_of_people_served` in our database. It is always important to think about data. We tend to just analyse, and calculate at the start of our careers, but the value we bring as data practitioners is in understanding the meaning of results or numbers, and interpreting their meaning. Imagine we were presenting this to the President and all of the Ministers, and one of them asks us: "Why does it say that 644 share a home tap?" and we had no answer.

This means that 1 tap_in_home actually represents 644 ÷ 6 = ‡ 100 taps. Calculating the average number of people served by a single instance of each water source type helps us understand the typical capacity or load on a single water source. This can help us decide which sources should be repaired or upgraded, based on the average impact of each upgrade. For example, wells don't seem to be a problem, as fewer people are sharing them. On the other hand, 2000 share a single public tap on average! We saw some of the queue times last time, and now we can see why. So looking at these results, we probably should focus on improving shared taps first.

Now let's calculate the total number of people served by each type of water source in total, to make it easier to interpret, order them so the most people served by a source is at the top.

In [86]:
%%sql 
SELECT type_of_water_source, ROUND(SUM(number_of_people_served)/(SELECT SUM(number_of_people_served) FROM water_source) *100) AS percentages
FROM water_source
GROUP BY type_of_water_source
ORDER BY percentages DESC;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('shared_tap', Decimal('43')),
 ('well', Decimal('18')),
 ('tap_in_home', Decimal('17')),
 ('tap_in_home_broken', Decimal('14')),
 ('river', Decimal('9'))]

43% of our people are using shared taps in their communities, and on average, we saw earlier, that 2000 people share one shared_tap. Pumps that serve these homes that are broken.

18% of people are using wells. But only 4916 out of 17383 are clean = 28%. By addin`g tap_in_ho`me an`d tap_in_home_brok`en together, we see that 31% of people have water infrastructure installed in their homes, but 45% (14/31) of these taps are not working! This isn't the tap itself that is broken, but rather the infrastructure like treatment plants, reservoirs, pipes, and pumps that serve these homes that are broken

## Start of a Solution

At some point, we will have to fix or improve all of the infrastructure, so we should start thinking about how we can make a data-driven decision how to do it. A simple approach is to fix the things that affect most people first. So let's write a query that ranks each type of source based on how many people in total use it.` RANK(`) should tell you we are going to need a window function to do this, so let's think through the problem.

In [96]:
%%sql
# Rank the types of water sources based on the number of people served
SELECT
    type_of_water_source,
    SUM(number_of_people_served) AS people_served,
    RANK() OVER(ORDER BY SUM(number_of_people_served) DESC) AS rank_by_population
FROM md_water_services.water_source
GROUP BY type_of_water_source;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('shared_tap', Decimal('11945272'), 1),
 ('well', Decimal('4841724'), 2),
 ('tap_in_home', Decimal('4678880'), 3),
 ('tap_in_home_broken', Decimal('3799720'), 4),
 ('river', Decimal('2362544'), 5)]

Ok, so we should fix shared taps first, then wells, and so on. But the next question is, which shared taps or wells should be fixed first? We can use the same logic; the most used sources should really be fixed first.

In [105]:
%%sql
# Rank the actual water sources based on the population served per source type
SELECT source_id, type_of_water_source, number_of_people_served,
DENSE_RANK() OVER(PARTITION BY type_of_water_source ORDER BY number_of_people_served DESC) AS priority_rank
FROM water_source
LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
10 rows affected.


[('SoRu36791224', 'river', 998, 1),
 ('SoRu36238224', 'river', 998, 1),
 ('SoRu36880224', 'river', 998, 1),
 ('SoRu37756224', 'river', 998, 1),
 ('SoRu38142224', 'river', 998, 1),
 ('KiRu30353224', 'river', 998, 1),
 ('KiZu31236224', 'river', 998, 1),
 ('KiRu28591224', 'river', 998, 1),
 ('KiMr25030224', 'river', 998, 1),
 ('SoRu35837224', 'river', 998, 1)]

By using RANK(), teams doing the repairs can use the value of rank to measure how many they have fixed, but what would be the benefits of using DENSE _RANK()? Maybe it is easier to explain to the engineers this way, or the priority feels a bit more natural?

## Analysing Queues

Recall that the `visits` table documented all of the visits the field surveyors made to each location. For most sources, one visit was enough, but if there were queues, they visited the location a couple of times to get a good idea of the time it took for people to queue for water. So we have the time that they collected the data, how many times the site was visited, and how long people had to queue for water. Remember we can use some `DateTime` functions here to get some deeper insight into the water queueing situation in Maji Ndogo, like which day of the week it was, and what time. Below are some questions to mull over:

How long did the survey tak
e?
What is the average total queue time for wat
er?
What is the average queue time on different d
ays?
How can we communicate this information efficie
ntly?
To answer the first question, we can run either of the func`tions DA`TEDI`FF or TIMESTA`MPDIFF (both will return the same output)

In [107]:
%%sql 
SELECT * FROM visits LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[(0, 'SoIl32582', 'SoIl32582224', datetime.datetime(2021, 1, 1, 9, 10), 1, 15, 12),
 (1, 'KiRu28935', 'KiRu28935224', datetime.datetime(2021, 1, 1, 9, 17), 1, 0, 46),
 (2, 'HaRu19752', 'HaRu19752224', datetime.datetime(2021, 1, 1, 9, 36), 1, 62, 40),
 (3, 'AkLu01628', 'AkLu01628224', datetime.datetime(2021, 1, 1, 9, 53), 1, 0, 1),
 (4, 'AkRu03357', 'AkRu03357224', datetime.datetime(2021, 1, 1, 10, 11), 1, 28, 14)]

In [106]:
%%sql
# Compute the duration of the survey in days
SELECT
    DATEDIFF(MAX(time_of_record), MIN(time_of_record)) AS "survey_duration (in days)"
    # TIMESTAMPDIFF(day, MIN(time_of_record), MAX(time_of_record)) AS "survey_duration (in days)"
FROM md_water_services.visits;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
1 rows affected.


[(924,)]

**924 days** which is equivalent to around two and a half years, that's how long the survey took. Let's turn our heads to the next question. Let's see how long people have to queue on average in Maji Ndogo. Keep in mind that many sources like `taps_in_home` have no queues. These are just recorded as `0` in the `time_in_queue` column, so when we calculate averages, we need to exclude those rows.

In [108]:
%%sql
# Compute the average waiting time in the queue
SELECT
    ROUND(AVG(NULLIF(time_in_queue, 0))) AS avg_waiting_time
FROM md_water_services.visits;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
1 rows affected.


[(Decimal('123'),)]

We get a queue time of about 123 min. So on average, people take two hours to fetch water if they don't have a tap in their homes. That may sound reasonable, but some days might have more people who need water, and only have time to go and collect some on certain days. Let's go a little bit further and try to answer the third question. Let's look at the queue times aggregated across the different times of the week.

`
DAY`() gives you the day of the month. If we want to aggregate data for each day of the week, we need to use another DateTime function`, DAYNAME(colum`n). As the name suggests, it returns the day from a timestamp as a string. Using that on th`e time_of_reco`rd column will result in a column with day names, Monday, Tuesday, etc., from the timestamp. To do this, we need to calculate the average queue time, grouped by day of the week.

In [14]:
%%sql
# Compute the average queue times per day of the week
SELECT 
    DAYNAME(time_of_record) AS day_of_the_week, 
    ROUND(AVG(NULLIF( time_in_queue, 0))) AS avg_waiting_time
FROM visits
GROUP BY day_of_the_week
ORDER BY avg_waiting_time DESC;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
7 rows affected.


[('Saturday', Decimal('246')),
 ('Monday', Decimal('137')),
 ('Friday', Decimal('120')),
 ('Tuesday', Decimal('108')),
 ('Thursday', Decimal('105')),
 ('Wednesday', Decimal('97')),
 ('Sunday', Decimal('82'))]

Alas! Saturdays have much longer queue times compared to the other days! That's interesting, Now let's go deeper and try to answer the fourth question to investigate what time during the day people collect water.

In [26]:
%%sql
SELECT 
    TIME_FORMAT(TIME(time_of_record), "%H:00") AS time,
    ROUND(AVG(NULLIF(time_in_queue, 0)))AS avg_waiting_time
FROM visits
GROUP BY time
ORDER BY avg_waiting_time DESC;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
14 rows affected.


[('19:00', Decimal('168')),
 ('07:00', Decimal('149')),
 ('08:00', Decimal('149')),
 ('17:00', Decimal('149')),
 ('06:00', Decimal('149')),
 ('18:00', Decimal('147')),
 ('09:00', Decimal('118')),
 ('13:00', Decimal('115')),
 ('10:00', Decimal('114')),
 ('14:00', Decimal('114')),
 ('15:00', Decimal('114')),
 ('16:00', Decimal('114')),
 ('12:00', Decimal('112')),
 ('11:00', Decimal('111'))]

We can deduce that mornings and evenings are the busiest from query result above. What could this mean? Are people collecting water before and after work? Wouldn't it be nice to break down the queue times for each hour of each day? In a spreadsheet, we can just create a pivot table.

Pivot tables are not widely used in SQL, despite being useful for interpreting results hence there are no built-in functions to do this for us. However sometimes the dataset is just overwhelmingly massive that it is the only option.

For rows, we will use the hour of the day in that nice format, and then make each column a different day! To filter a row we use WHERE, but us`ing CA`SE() in SELECT can filter columns. We can us`e a CA`SE() function for each day to separate the queue time column into a column for each day. Let's begin by only focusing` on Su`nday. So, when a ro`w's DAYNAME(time_of_rec`ord) is Sunday, we make that value equal` to time_in_q`ueue, `and `NULL for any other days.

In [32]:
%%sql
# Create a pivot table of avg queue times per hour of day across days of the week
SELECT
	TIME_FORMAT(TIME(time_of_record), "%H:00") AS hour_of_day,
    # Sunday
    ROUND(AVG(
		CASE 
			WHEN DAYNAME(time_of_record) = "Sunday"
            THEN time_in_queue ELSE null
		END), 0) AS Sunday,
	
    # Monday
    ROUND(AVG(
		CASE 
			WHEN DAYNAME(time_of_record) = "Monday" 
            THEN time_in_queue ELSE null
		END), 0) AS Monday,
	
    # Tuesday
    ROUND(AVG(
		CASE 
			WHEN DAYNAME(time_of_record) = "Tuesday" 
            THEN time_in_queue ELSE null
		END), 0) AS Tuesday,
	
    # Wednesday
    ROUND(AVG(
		CASE
			WHEN DAYNAME(time_of_record) = "Wednesday" 
            THEN time_in_queue ELSE null
		END), 0) AS Wednesday,
	
    # Thursday
    ROUND(AVG(
		CASE
			WHEN DAYNAME(time_of_record) = "Thursday" 
            THEN time_in_queue ELSE null
		END), 0) AS Thursday,
    
    # Friday
	ROUND(AVG(
		CASE
			WHEN DAYNAME(time_of_record) = "Friday" 
            THEN time_in_queue ELSE null
		END), 0) AS Friday,
	
    # Saturday
    ROUND(AVG(
		CASE
			WHEN DAYNAME(time_of_record) = "Saturday" 
            THEN time_in_queue ELSE null
		END), 0) AS Saturday
FROM md_water_services.visits
WHERE time_in_queue != 0
GROUP BY hour_of_day
ORDER BY hour_of_day;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
14 rows affected.


[('06:00', Decimal('79'), Decimal('190'), Decimal('134'), Decimal('112'), Decimal('134'), Decimal('153'), Decimal('247')),
 ('07:00', Decimal('82'), Decimal('186'), Decimal('128'), Decimal('111'), Decimal('139'), Decimal('156'), Decimal('247')),
 ('08:00', Decimal('86'), Decimal('183'), Decimal('130'), Decimal('119'), Decimal('129'), Decimal('153'), Decimal('247')),
 ('09:00', Decimal('84'), Decimal('127'), Decimal('105'), Decimal('94'), Decimal('99'), Decimal('107'), Decimal('252')),
 ('10:00', Decimal('83'), Decimal('119'), Decimal('99'), Decimal('89'), Decimal('95'), Decimal('112'), Decimal('259')),
 ('11:00', Decimal('78'), Decimal('115'), Decimal('102'), Decimal('86'), Decimal('99'), Decimal('104'), Decimal('236')),
 ('12:00', Decimal('78'), Decimal('115'), Decimal('97'), Decimal('88'), Decimal('96'), Decimal('109'), Decimal('239')),
 ('13:00', Decimal('81'), Decimal('122'), Decimal('97'), Decimal('98'), Decimal('101'), Decimal('115'), Decimal('242')),
 ('14:00', Decimal('83'), Decimal('127'), Decimal('104'), Decimal('92'), Decimal('96'), Decimal('110'), Decimal('244')),
 ('15:00', Decimal('83'), Decimal('126'), Decimal('104'), Decimal('88'), Decimal('92'), Decimal('110'), Decimal('248')),
 ('16:00', Decimal('83'), Decimal('127'), Decimal('99'), Decimal('90'), Decimal('99'), Decimal('109'), Decimal('251')),
 ('17:00', Decimal('79'), Decimal('181'), Decimal('135'), Decimal('121'), Decimal('129'), Decimal('151'), Decimal('251')),
 ('18:00', Decimal('80'), Decimal('174'), Decimal('122'), Decimal('113'), Decimal('132'), Decimal('158'), Decimal('240')),
 ('19:00', Decimal('127'), Decimal('159'), Decimal('145'), Decimal('176'), Decimal('137'), Decimal('103'), Decimal('282'))]

Perfect 👌🏾, now we can compare the queue times for each day, hour by hour! The following patterns emerge with careful observation of the custom built SQL pivot table 😎:

1. Queues are very long on a Monday morning and Monday evening as people rush to get water.
2. Wednesday has the lowest queue times (on weekdays), but long queues on Wednesday evening.
3. People have to queue pretty much twice as long on Saturdays compared to the weekdays. It looks like people spend their Saturdays queueing for water, perhaps for the week's supply?
4. The shortest queues are on Sundays, and this is probably a cultural thing. The people of Maji Ndogo may prioritise family and religion, so Sundays are spent with family and friends.

SQL is a set of tools we can apply. By understanding CASE() function, we could build a complex query that aggregates our data in a format that is very easy to understand.

# Water Accessibility and infrastructure summary report

This survey aimed to identify the water sources people use and determine both the total and average number of users for each source. Additionally, it examined the duration citizens typically spend in queues to access water. So let's create a short summary report we can send off to relevant stakeholders:ds.

## Insights

1. Most water sources are rural.
2. 43% of our people are using shared taps. 2000 people often share one tap.
3. 31% of our population has water infrastructure in their homes, but within that group 45% face non-functional systems due to issues with pipes, pumps, and reservoirs.
4. 18% of our people are using wells of which, but within that, only 28% are clean.
5. Our citizens often face long wait times for water, averaging more than 120 minutes.

es.
In terms of qu queues.


- Queues are very long on Saturdays.
- 
Queues are longer in the mornings and evenings
- 
Wednesdays and Sundays have the shortest queu.eseues:

# Strategy to Solving the Water Crisis in Maji Ndogo

If we were to think about a strategy to start alleviating Maji Ndogo’s water crisis, it would be something like this:

1. We want to focus Maji Ndogo’s efforts on improving water sources that affect the most people.

   - Most people will benefit if we improve the shared taps first.
   - Wells are a good source of water, but many are contaminated. Fixing this will benefit a lot of people.
   - Fixing existing infrastructure will help many people. If they have running water again, they won’t have to queue, thereby shortening queue times for others *Killing two birds with one stone.*
   - Installing taps in homes will stretch Maji Ndogo’s resources too thin, so for now, if the queue times are low, we won’t improve that source.

2. Most water sources are in rural areas. We need to ensure the teams know this as this means they will have to make these repairs/upgrades in rural areas where road conditions, supplies, and labour are harder challenges to overcome.
o overcome.

## Recommended Practical Solutions

1. If communities are using **rivers**, we can dispatch trucks to those regions to provide water temporarily in the short term, while we send out crews to drill for wells, providing a more permanent solution.

2. If communities are using **wells**, we can install filters to purify the water. For wells with **biological** contamination, we can **install UV filters** that kill microorganisms, and for **chemically** polluted wells, we can **install reverse osmosis (RO) filters**. In the long term, we need to figure out why these sources are polluted.

3. For **shared taps**, in the short term, we can send additional water tankers to the busiest taps, on the busiest days. We can use the queue time pivot table we made to send tankers at the busiest times. Meanwhile, we can start the work on **installing extra taps** where they are needed. According to UN standards, the maximum acceptable wait time for water is 30 minutes. With this in mind, our aim is to **install taps to get queue times below 30 min.**

4. **Shared taps with short queue times** (< 30 min) represent a logistical challenge to further reduce waiting times. The most effective solution, installing taps in homes, is resource-intensive and better suited as a long-term goal.

5. **Addressing broken infrastructure** offers a significant impact even with just a single intervention. It is expensive to fix, but so **many people can benefit** from repairing one facility. For example, fixing a reservoir or pipe that multiple taps are connected to. We will have to find the commonly affected areas though to see where the problem actually is.